test/apps/iodemo/io_demo.cc

```diff
@@ -832,132 +889,182 @@ class DemoClient : public P2pDemoCommon {
+        }
+    }
+
+    void connect_all() {
+        for (int server_index = 0; server_index < _server_info.size();
+             ++server_index) {
+            server_info_t& server_info = _server_info[server_index];
+            if ((server_info.conn != NULL) ||
+                (server_info.retry_count >= opts().client_retries)) {
+                /* server is already connected, or retry count exceeded */
+                continue;
             }
+
+            server_info.conn = connect(opts().servers[server_index]);
+            if (server_info.conn == NULL) {
+                connect_failed(server_index);
+                continue;
+            }
+
+            server_info.retry_count = 0;
+            _conn_idx[server_info.conn] = server_index;
+
+            server_info.active_index = _active_servers.size();
+            _active_servers.push_back(server_index);
+
+            LOG << "Connected to " << server_name(server_index);
         }
+    }
 
-        // reset number of retries after successful connection
-        _retry = 0;
+    status_t run() {
+        _server_info.resize(opts().servers.size());
+        std::for_each(_server_info.begin(), _server_info.end(),
+                      reset_server_info);
 
         _status = OK;
 
         // TODO reset these values by canceling requests
         _num_sent      = 0;
         _num_completed = 0;
 
-        uint32_t sn          = IoDemoRandom::rand<uint32_t>();
-        double prev_time     = get_time();
-        long total_iter      = 0;
-        long total_prev_iter = 0;
-        std::vector<op_info_t> info;
-
-        for (int i = 0; i < IO_OP_MAX; ++i) {
-            op_info_t op_info = {static_cast<io_op_t>(i), 0, 0};
-            info.push_back(op_info);
-        }
+        uint32_t sn                  = IoDemoRandom::rand<uint32_t>();
+        double prev_time             = get_time();
+        double prev_connect_time     = 0;
+        long total_iter              = 0;
+        long total_prev_iter         = 0;
+        op_info_t op_info[IO_OP_MAX] = {{0,0}};
 
         while ((total_iter < opts().iter_count) && (_status == OK)) {
             VERBOSE_LOG << " <<<< iteration " << total_iter << " >>>>";
 
-            if (!wait_for_responses(opts().window_size - 1)) {
+            wait_for_responses(opts().window_size - 1);
+            if (_status != OK) {
                 break;
             }
 
-            size_t conn_num = IoDemoRandom::rand(size_t(0), conn.size() - 1);
-            io_op_t op      = get_op();
+            /* Always try to reconnect every if all servers are disconnected,
+             * or try every 10 iterations if only some servers are disconnected
+             */
+            if ((_active_servers.size() < _server_info.size()) &&
+                (_active_servers.empty() || ((total_iter % 10) == 0))) {
+                double curr_time = get_time();
+                if (curr_time > (prev_connect_time + opts().client_timeout)) {
```

Comment:
maybe add separate function for this check?

Reply:
'total_iter' and 'prev_connect_time' are needed from the calling context, IMO it would be ugnlier

Reply:
make _prev_connect_time a member

Reply:
Replaced prev_connect_time with member variable _prev_connect_time.